test/cpp/end2end/xds_end2end_test.cc

```diff
@@ -4348,48 +4347,48 @@ TEST_P(LdsRdsTest, XdsRoutingApplyXdsTimeout) {
     balancers_[0]->ads_service()->SetLdsResource(listener);
   }
   // Test grpc_timeout_header_max of 1.5 seconds applied
-  auto t0 = system_clock::now();
+  gpr_timespec tolerance = gpr_time_from_millis(kToleranceMillis, GPR_TIMESPAN);
+  gpr_timespec est_timeout_time = gpr_time_add(
+      gpr_now(GPR_CLOCK_REALTIME),
+      gpr_time_from_millis(
+          kTimeoutGrpcTimeoutHeaderMaxSecond * 1000 + kTimeoutMillis,
+          GPR_TIMESPAN));
   CheckRpcSendFailure(1,
                       RpcOptions()
                           .set_rpc_service(SERVICE_ECHO1)
                           .set_rpc_method(METHOD_ECHO1)
                           .set_wait_for_ready(true)
                           .set_timeout_ms(kTimeoutApplicationSecond * 1000),
                       StatusCode::DEADLINE_EXCEEDED);
-  auto ellapsed_nano_seconds =
-      std::chrono::duration_cast<std::chrono::nanoseconds>(system_clock::now() -
-                                                           t0);
-  EXPECT_GT(ellapsed_nano_seconds.count(),
-            kTimeoutGrpcTimeoutHeaderMaxSecond * 1000000000 + kTimeoutNano);
-  EXPECT_LT(ellapsed_nano_seconds.count(),
-            kTimeoutMaxStreamDurationSecond * 1000000000);
+  gpr_timespec timeout_time = gpr_now(GPR_CLOCK_REALTIME);
+  GPR_ASSERT(gpr_time_similar(est_timeout_time, timeout_time, tolerance));
```

Comment:
This allows the two times to be off in either direction.  I think that's more permissive than what we want for this test, and I'm concerned that this may be masking the real problem, considering that the failure mode we saw previously was that the timeout happened too early.  I think the test should allow the timeout to take a little longer, but I don't think it should ever happen sooner than expected.  The test should fail if that occurs.